### DISCUSSION TITLE
help needed to highlight parts of a formula when I hover over the part

### DISCUSSION THREAD

**Discussion by fbeleznay** [KaTeX/KaTeX#3764] [Q&A]
I am trying to highlight parts of a formula when I hover over the part.
I tried different ways, but only partially succeeded (my code is copied at the end of this post).
My code works with a usual looking formula, but have problems when I add a box around the parts that I would like to select.

- In the first formula the 0 (in the lower limit of the integral) is highlighted fine when I move the cursor close enough to the 0.

- In the second formula there is a very limited region for the cursor for the highlighting to appear. It seems that the box hides the content for the html style.

- In the third formula the highlighting works when the cursor is above the border of the box, but not everywhere inside. Even then, the background highlighting does not fill the full box.

- In the fourth example I used nested boxes. The partial highlighting works for the outer box, but now the highlighting stays active when the cursor is inside the box.

My ultimate goal would be something that highlights the full content of the outer box when the cursor is on the border (or inside) and only highlights the inner box when the cursor is on the border (or inside) this inner box.
Any help on achieving this goal would be appreciated.

```
<!DOCTYPE html>
<html>
    <head>

        <link rel="stylesheet" <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.css" integrity="sha384-Xi8rHCmBmhbuyyhbI88391ZKP2dmfnOl4rT9ZfRI7mLTdk1wblIUnrIq35nqwEvC" crossorigin="anonymous">
        <script src="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.js" integrity="sha384-X/XCfMm41VSsqRNQgDerQczD69XqmjOOOwYQvr/uuC+j4OPoNhVgjdGFwhvN02Ja" crossorigin="anonymous"></script>
        <script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/contrib/auto-render.min.js" integrity="sha384-+XBljXPPiv+OzfbB3cVmLHf4hdUFHlWNZN5spNQ7rmHTXpd7WvJum6fIACpNNfIR" crossorigin="anonymous"
        onload="renderMathInElement(document.body);"></script>
        
        <meta name="viewport" content="width=device-width, initial-scale=1">
        
        <style>
            .parts:hover {background-color: red;}
        </style>
    </head>
    <body>
        <p id="expression1">
            ???
        </p>
        <p id="expression2">
            ???
        </p>
        <p id="expression3">
            ???
        </p>
        <p id="expression4">
            ???
        </p>
        
        <script>
            var x = "\\displaystyle\\int_{\\htmlClass{parts}{0}}^{1}";
            var w1 = document.getElementById("expression1");
            katex.render(x,w1, {strict: false, trust: true})
            x = "\\displaystyle\\int_{\\boxed{\\htmlClass{parts}{0}}}^{1}";
            var w2 = document.getElementById("expression2");
            katex.render(x,w2, {strict: false, trust: true})
            x = "\\displaystyle\\int_{\\htmlClass{parts}{\\boxed{0}}}^{1} ";
            var w3 = document.getElementById("expression3");
            katex.render(x,w3, {strict: false, trust: true})
            x = "\\htmlClass{parts}{\\boxed{\\displaystyle\\int_{\\htmlClass{parts}{\\boxed{0}}}^{1}}}";
            var w4 = document.getElementById("expression4");
            katex.render(x,w4, {strict: false, trust: true})            
        </script>

    </body>
</html>
```

**Comment by mbourne** [comment]
Hello fbeleznay

I suspect what you're aiming to do within an integral expression like this is not going to be possible with "native" KaTeX. (I'm happy to be proved wrong).

But here's a workaround where I've created a new "expression5" based on your "expression4".

Firstly, I'm going to remove the `\box` element in your math expression:

```x = "\\htmlClass{parts}{{\\displaystyle\\int_{\\htmlClass{parts}{{0}}}^{1}}}";```

I'm going to create my own box since (as you've found) the "native" box is placed behind the math, so the hover doesn't work as desired.

Here's the outer box (based on the "base" class)

```#expression5 .base{padding:0.275em 0.275em 0.4em;border:1px solid #911;user-select:none}```

Here's the hover on the outer box:

```#expression5 .base:hover{background-color: #f77}```

![image](https://user-images.githubusercontent.com/7541940/209867916-6ad2d3d8-b0dc-4703-b1b0-d1ffa0d71bb3.png)

The box around the lower limit:

```#expression5 .parts .parts{padding:0.25em;border:1px solid #933;background:#fff}```

The hover for the lower limit:

```#expression5 .parts .parts:hover{background:#f77}```

There's a vlist element above the right part of our lower box that stops the hover working on the whole box, so I remove it as follows:

```#expression5 .vlist-t2 .vlist-r:last-child{display:none}```

Normally the sky would fall in if you do something like this (the spacing normally goes all over the place) but, here, it seems OK.

![image](https://user-images.githubusercontent.com/7541940/209868130-20c76a96-189d-4080-8c9d-cfcee5ab1411.png)

Finally, to tidy up, I added this so the global "parts" hover didn't mess with my "expression5":

```#expression5 .parts:hover{background:#f77}```

Below is the revised page.

Hope it helps

Regards
Murray

```
<!DOCTYPE html>
<html>
    <head>

        <link rel="stylesheet" <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.css" integrity="sha384-Xi8rHCmBmhbuyyhbI88391ZKP2dmfnOl4rT9ZfRI7mLTdk1wblIUnrIq35nqwEvC" crossorigin="anonymous">
        <script src="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.js" integrity="sha384-X/XCfMm41VSsqRNQgDerQczD69XqmjOOOwYQvr/uuC+j4OPoNhVgjdGFwhvN02Ja" crossorigin="anonymous"></script>
        <script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/contrib/auto-render.min.js" integrity="sha384-+XBljXPPiv+OzfbB3cVmLHf4hdUFHlWNZN5spNQ7rmHTXpd7WvJum6fIACpNNfIR" crossorigin="anonymous"
        onload="renderMathInElement(document.body);"></script>
        
        <meta name="viewport" content="width=device-width, initial-scale=1">
        
        <style>
			body{font-size:3em}		
			.parts:hover {background-color: red;}
			
			#expression5 .parts:hover{background:#f77}
			#expression5 .base{padding:0.275em 0.33em 0.4em;border:1px solid #911;user-select:none}
			#expression5 .base:hover{background-color: #f77}
			#expression5 .parts .parts{padding:0.15em 0.25em;border:1px solid #933;background:#fff}
			#expression5 .parts .parts:hover{background:#f77}
			
			/* Uncomment this to see all the spans involved: */
			/* #expression5 span{border:1px solid #f0f} */
			
			#expression5 .vlist-t2 .vlist-r:last-child{display:none}
			
        </style>
    </head>
    <body>
        <p id="expression1">
            ???
        </p>
        <p id="expression2">
            ???
        </p>
        <p id="expression3">
            ???
        </p>
        <p id="expression4">
            ???
        </p>
		<p id="expression5">
            ???
        </p>
        
        <script>
            var x = "\\displaystyle\\int_{\\htmlClass{parts}{0}}^{1}";
            var w1 = document.getElementById("expression1");
            katex.render(x,w1, {strict: false, trust: true})
            x = "\\displaystyle\\int_{\\boxed{\\htmlClass{parts}{0}}}^{1}";
            var w2 = document.getElementById("expression2");
            katex.render(x,w2, {strict: false, trust: true})
            x = "\\displaystyle\\int_{\\htmlClass{parts}{\\boxed{0}}}^{1} ";
            var w3 = document.getElementById("expression3");
            katex.render(x,w3, {strict: false, trust: true})
            x = "\\htmlClass{parts}{\\boxed{\\displaystyle\\int_{\\htmlClass{parts}{\\boxed{0}}}^{1}}}";
            var w4 = document.getElementById("expression4");
            katex.render(x,w4, {strict: false, trust: true})	
			
			x = "\\htmlClass{parts}{{\\displaystyle\\int_{\\htmlClass{parts}{{0}}}^{1}}}";
			var w5 = document.getElementById("expression5");
            katex.render(x,w5, {strict: false, trust: true}) 
        </script>

    </body>
</html>
```

**Reply by fbeleznay** [reply]
Thank you, this is very helpful. I do not quite understand the code, so I need to look at the details so that I can modify it to fit my purpose. I am trying to create an equation editor that knows the supported functions of KaTeX. I currently have the following: [belez.info/BF/KaTeX/EquationEditor.html](http://belez.info/BF/KaTeX/EquationEditor.html). I am trying to add a graphical part where boxes are used to select and edit different parts. I know that there are other online equation editors, so I am not quite sure why I am ding this, but nevertheless I still did not give up. These are the points that I need to understand in your code to modify:

- I need the highlighting the other way around (highlight everything when the cursor is in the large box, but not in the small box and highlight **only** the small box when the cursor is inside).
- I need nested boxes with unlimited number of layers.
- The formula is written by the user, so I need to generate the objects depending on what the user is typing.
- I need to act on click, not only highlighting on hover.

Probably I will need some javascript to code the underlining structure of the formula first and use KaTeX only to typeset the output, but I thought I try first the HTML option of KaTeX to identify parts of the expression.

Thanks again for your help.

**Comment by mbourne** [comment]
> " I do not quite understand the code"

When I don't know what someone's CSS is doing, I comment out the whole thing, then add one line at a time to see what it does.

> "I need the highlighting the other way around"

The revised page (I believe) now does what you want for `#expression5`. I couldn't see a way to do it with CSS only, so it's javascript-based. (I've removed the earlier 4 expressions and the generic "parts" class of your page.)

I would suggest though, that you get all the other aspects of your interface worked out before worrying about highlighting.

> "I am trying to create an equation editor"

Since more than half of the users of the Web are now on small screens, could I strongly encourage you to develop for mobile first, then ensure it also works on larger screens.

One way you could go about an editor is to make the various "boxes" of existing kaTeX expressions editable. 

In the revised page, I add a "contenteditable" attribute to the limits, the integrand and the differential of the integral. The user can type in the (green border) boxes. As they type, the math rendering is updated. You can do things like this:

![image](https://user-images.githubusercontent.com/7541940/210022726-64eb818e-d6a0-4ac0-baf9-df8476105918.png)

There's no error handling yet (you'll see a console error when beginning to type things like \sin or \pi) but I'll leave that as an exercise for the reader. :-) Alternatively, you could add a "click" listener to your button so the output only updates when the user is ready.

Once again, I hope it helps get you started.
Regards
Murray

### New style:
```
<style>
	body p{font-size:1.5em}

	#expression5 .base{padding:0.275em 0.33em 0.4em;border:1px solid #911;user-select:none}
	#expression5 .base:hover{background-color: #f77}

	.partsInner{padding:0.15em 0.25em;border:1px solid #933}
	.partsInner:hover{background:#f77}
	.vlist-t2 .vlist-r:last-child{display:none}

	.editMe{padding:0.15em 0.25em;border:1px solid #393;font-family:monospace}
	.editMe .mathnormal{font-family:monospace;font-style:normal}
	.editMe:hover{background:#7f7}
</style>
```
### New HTML
```
<h1>Highlight and Editable</h1>
<h2>Highlight</h2>
<p id="expression5">
	???
</p>
<h2>Editable Integral Expression</h2>
<p id="expression6">
	???
</p>
<h2>Output</h2>
<p id="exp6_output">
	???
</p>
```
### New script
```
<script>

	x = "\\displaystyle\\int_{\\htmlClass{partsInner}{{0}}}^{1}";
	var w5 = document.getElementById("expression5");
	katex.render(x,w5, {strict: false, trust: true})
	
	// Achieve desired hover effects on #expression5

	document.querySelector('#expression5 .partsInner').addEventListener('mouseover', function() {
		document.querySelector('#expression5 .base').style.background = "#fff";
	});

	document.querySelector('#expression5 .partsInner').addEventListener('mouseout', function() {
		document.querySelector('#expression5 .base').style.background = null;
	});
	
	// In #expression6, add "editMe" class to limits, integrand and differential

	x = "\\displaystyle\\int_{\\htmlClass{editMe}{{0}}}^{\\htmlClass{editMe}{1}}{\\htmlClass{editMe}{x}}{\\htmlClass{editMe}{dx}}";
	var w6 = document.getElementById("expression6");
	katex.render(x,w6, {strict: false, trust: true})

	var ems = document.querySelectorAll('.editMe');
	
	// Make each "editMe" editable	

	ems.forEach(function(em, i){
		em.id = "exp6_em_"+i;
		em.setAttribute("contenteditable", "true");
		em.addEventListener("keyup", updateOutput);
	});
	
	// Function for updating the output expression as user types

	function updateOutput() {
		var lower = document.querySelector("#exp6_em_0").textContent;
		var upper = document.querySelector("#exp6_em_1").textContent;
		var integrand = document.querySelector("#exp6_em_2").textContent;
		var differential = document.querySelector("#exp6_em_3").textContent;

		x = "\\displaystyle\\int_{" + lower + "}^{" + upper +"}"+integrand+differential;
		var w6_output = document.getElementById("exp6_output");
		katex.render(x,w6_output, {strict: false, trust: true})
	}
	updateOutput();

</script>
```
